### DISCUSSION TITLE
모니터링 대시보드 어떤 게 좋을까?

### DISCUSSION THREAD

**Discussion by donghoony** [woowacourse-teams/2024-review-me#204] [BE 이야깃거리]
이건 뭘까
로그와 모니터링 도구 학습하고 서로 알려주기

산초: Grafana + Loki
커비: CloudWatch, 모니터링 필요성
테드: Prometheus
아루: ELK + Actuator (가능하면)

**Comment by donghoony** [comment]
# ELK Stack
- ElasticSearch, Logstash, Kibana 세 가지를 합쳐서 ELK라고 부른다. 
- 각각 검색엔진, 수집, 시계열 데이터 시각화를 담당한다.
- ___Beats: 데이터를 수집하는 데에 도움을 주는 도구. FileBeat는 로그 파일을, HeartBeat는 가동 시간 모니터링, PacketBeat은 네트워크 데이터 등이다.

# Actuator
스프링 내부의 데이터를 모아줌, 메모리에 저장되기 때문에 추가적인 데이터 모으기가 필요
`spring-boot-starter-actuator`

**Reply by donghoony** [reply]
ElasticSearch / OpenSearch

Logstash

시계열 데이터 수집
Prometheus

**Reply by donghoony** [reply]
데이터 시각화
- Grafana
- Kibana
- Datadog (유료, 무료버전은 로그 저장 1일)
- Cerebro (업데이트 안한 지 오래 됐음, 2019~)
- Prometheus

**Reply by donghoony** [reply]
https://hudi.blog/elk-stack-overview/

**Reply by donghoony** [reply]
https://ssyoni.tistory.com/33

**Reply by donghoony** [reply]
https://docs.spring.io/spring-boot/reference/features/logging.html#features.logging.custom-log-configuration

**Reply by donghoony** [reply]
전체 플로우: Actuator -> Logback -> LogStash -> ElasticSearch -> Kibana

**Comment by Kimprodp** [comment]
## Prometheus
- 메트릭 기반 모니터링에 최적화된 도구로 메트릭을 받아서 저장한다.
- 때문에 이벤트 로그나 개별 이벤트를 저장하는 일에는 적합하지 않다.
- 다른 모니터링 툴과 다르게 `pull` 방식을 사용하여 메트릭을 수집한다.
   > **push 방식의 문제점**
  push 방식은 모니터링 대상 서버에서 부하가 높은 상황일 때 메트릭 정보를 전송하고자 한다면 이는 어플리케이션에 문제를 발생시킬 수 있는 포인트가 될 수 있다. 반면 pull 방식은  모니터링을 하는 주체가 원할 때만 가져갈 수 있도록 할 수 있어 대상 서버의 부하 부담을 줄여줄 수 있다.
- 자체적인 모니터링 GUI를 제공하며, 데이터를 분석하기 위한 PromQL(고유 쿼리 언어)를 통해 데이터를 집계하고 GUI에 노출 시킬 수 있다.
- Promethus의 모니터링 GUI는 시안성이 나쁘기 때문에 Grafana 대시보드를 연동하여 사용하는 방식이 많이 선호된다.
- Alert 시스템이 존재하여 PromQL을 통해 조건을 정의하고 email, slack을 알림 수단으로 사용할 수 있다.

**Comment by nayonsoso** [comment]
https://celestial-sheep-6b7.notion.site/Grafana-Loki-8e59997202904fac84f90b0f57ef045f

핵심만 보려면 맨 아래를 보시오!!

**Comment by hyeonji1220** [comment]
# 모니터링이 왜 필요하지?

### 1. **성능 모니터링**

> 해피케이스일 때 확인 목적. 문제가 터지진 않았지만 잘 돌아가는지!
> 
- 리뷰 작성, 조회 요청에 대한 평균 응답 시간 모니터링
    - 이 정보를 통해 서비스의 성능 저하를 신속하게 감지
- **리소스 사용량**
    - 서버의 CPU, 메모리, 네트워크 사용량을 모니터링하여 자원 부족으로 인한 성능 저하를 방지합니다.
- **데이터베이스 용량**
    - 리뷰 데이터의 저장 용량을 모니터링하여 데이터베이스 용량이 부족해지지 않도록 감시
- **쿼리 성능**
    - db 서버의 요청, 응답시간을 모니터링해서 너무 오래걸리면
    - 특정 api 요청시, 또는 리뷰 조회에 너무 많은 쿼리가 나가가나, 조인이 많거나를 확인함

### 2. **에러 모니터링**

> 문제 터졌을 때 확인 목적
> 
- 특정 유형의 에러가 증가하는 경우 파악
    - 실패한 리뷰 작성 및 조회 요청의 비율을 모니터링 → 특정 유형의 에러가 증가하는 경우, 즉시 문제를 진단하고 해결할 수 있습니다.
- **로그 분석**
    - 에러 로그와 서버 로그를 주기적으로 분석하여 문제의 원인을 파악합니다.

### 3. **사용자 행동 분석**

> 좋은 사용자 경험을 위해 서비스 운영에 반영하기 위해서
> 
- **작성 빈도**
    - 특정 화면에서 작성 빈도가 감소한다면 여기서 사용자가 떠나는 구나! 를 알 수 있음
- **조회 빈도**
    - 리뷰 조회 패턴을 분석
    - 상세 리뷰 조회를 생각보다 금방 보고 나간다던지

## 그럼 어떤 정도로 설정해놓으면 좋을까?

- 설정이 어렵지 않으면 다 설정해놓으면 좋지! 기록이야 나중에 필요할 수도 있으니까. 나중에 안쓰면 지우면 되고.
- 또는 필요하다고 생각되는 것만 설정하고, 나중에 일이 닥치고 아 이거 설정 안해놨지 필요한 데이터인데..를 느끼고 설정해도 되겠다.

# EC2  Cloud Watch 기본 제공 데이터
> EC2 Cloud Watch 에서 별도 설정 없이 확인 가능하다는 장점!

1. CPU 사용률(CPUUtilization)
    - - EC2 인스턴스의 CPU가 사용되고 있는 비율을 측정 
    - 예를 들어, CPU 사용률이 70%라면 인스턴스의 CPU 자원의 70%가 현재 사용되고 있다는 의미
2. 네트워크 사용률(NetworkIn, NetworkOut)
    - 인스턴스가 수신한 총 데이터 양 나타냄
3. 디스크 성능(DiskReadOps, DiskWriteOps)
    - 디스크에서 읽기 작업이 수행된 횟수
4. 디스크 읽기/쓰기(DiskReadBytes, DiskWriteBytes)
    - EC2 인스턴스가 디스크에서 데이터를 읽을때, 쓸 때 데이터의 총 양을 측정합니다.

## 미제공
메모리, 디스크 여유 공간
    - Cloudwatch Agent를 인스턴스에 설치하면 이런 추가 지표 확인 가능
    - 메모리 사용률, 디스크 스왑 사용률, 디스크 공간 사용률
